### PR TITLE
Spatial Distances

### LINKED ISSUE #9235: Implement scipy.spatial
It would be nice to implement `scipy.spatial`, starting from `distance` for pairwise distance computations.

### PR DESCRIPTION
**Pull request by ingmarschuster**
Initial implementation of
* distance primitives
* pdist

This would fix #9235 once done

### CODE REVIEW

File: jax/scipy/spatial/distance.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import jax.numpy as jnp
```
Comment (jakevdp) [inline jax/scipy/spatial/distance.py:1]: Top of file needs a copyright statement

File: jax/scipy/spatial/distance.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+from jax import jit, vmap
+
+# define primitives for distance functions according to scipy.spatial.distance
+# which can be batched over the first dimension and then used inside jit and vmap
+# as well as used for implementing cdist and pdist versions
+
+primitives = {
```
Comment (jakevdp) [inline jax/scipy/spatial/distance.py:8]: Rather than a dict of anonymous functions, I'd probably define these as named functions in the `distance` namespace, like `scipy.spatial.distance.braycurtis`, etc. It will make the implementations easier to read.

If it's useful for `pdist` you could then refer to these implementations in a dict.

Also, I'd validate the input shapes as well – the implementations here will return strange results if you pass in broadcast-compatible multi-dimensional arrays

File: jax/scipy/spatial/distance.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+# vectorize primitives for batched, jit'ed evaluation. Take special care for the primitives that
+# take additional arguments (p, V)
+_METRICS = {}
+for k, v in primitives.items():
+    if k in ("minkowski", "mahalanobis", "seuclidean"):
+        _METRICS[k] = jit(vmap(v, in_axes=(0, 0, None)))
+    else:
+        _METRICS[k] = jit(vmap(v))
```
Comment (jakevdp) [inline jax/scipy/spatial/distance.py:53]: I'd do the `jit` & `vmap` at the call site. It's clearer and shouldn't change performance characteristics appreciably.

File: jax/scipy/spatial/distance.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+
+def pdist(X, metric="euclidean", p=2, V=None):
```
Comment (jakevdp) [inline jax/scipy/spatial/distance.py:56]: The scipy function allows general `**kwargs` to be forwarded to the distance metric – that would prevent the need for special cases below.

File: jax/scipy/spatial/distance.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+# vectorize primitives for batched, jit'ed evaluation. Take special care for the primitives that
+# take additional arguments (p, V)
+_METRICS = {}
+for k, v in primitives.items():
+    if k in ("minkowski", "mahalanobis", "seuclidean"):
+        _METRICS[k] = jit(vmap(v, in_axes=(0, 0, None)))
+    else:
+        _METRICS[k] = jit(vmap(v))
+
+
```
Comment (jakevdp) [inline jax/scipy/spatial/distance.py:55]: You can use `@_wraps(scipy.spatial.distance.pdist)` and avoid defining your own doc string. Look at the functions in e.g. `jax.scipy.special` to see how we typically approach this.

File: jax/scipy/spatial/distance.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    >>> pdist(X, metric="rogerstanimoto")
+    array([0.33333333, 0.5       , 0.33333333])
+    >>> pdist(X, metric="russellrao")
+    array([0.5, 1. , 0.5])
+    >>> pdist(X, metric="sokalmichener")
+    array([0.33333333, 0.5       , 0.33333333])
+    >>> pdist(X, metric="sokalsneath")
+    array([0.33333333, 0.5       , 0.33333333])
+    """
+    if metric not in _METRICS:
```
Comment (jakevdp) [inline jax/scipy/spatial/distance.py:132]: This won't let callable metrics through

**Review by jakevdp — CHANGES_REQUESTED** [review]
This is a great start! A few specific comments below.

Some more general comments:

- we typically would put implementations in `jax/_src/scipy/spacial/distance.py`, and then import the public symbols in `jax/scipy/spatial/distance.py`. You can see other examples of this in the JAX codebase.
- this will need tests; feel free to create a new file like `tests/scipy_spatial_test.py` and you can follow other test files as a template.
- the new functions should be added to the docs in `docs/jax.scipy.rst`

Thanks for working on this – it's exciting to see this come together!

**Comment by jakevdp** [comment]
Hi - sorry that we lost track of this contribution. In the meantime we've had some discussion about the scope of the JAX numpy/scipy wrappers, and decided that distance metrics are out-of-scope for now. See https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html

Thanks!
